### PR TITLE
Client implementation

### PR DESCRIPTION
**Pull request by mqliang**
@thc202

First commit is the Client implementation

The second commit  is auto generated by https://github.com/zaproxy/zaproxy/pull/3670

cc @psiinon

### CODE REVIEW

File: zap/client.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		},
+	}
+	return &Client{
+		Config:     cfg,
+		httpClient: httpClient,
+	}, nil
+}
+
+func (c *Client) Request(path string, queryParams map[string]string) (interface{}, error) {
+	body, err := c.request(c.BaseOther+path, queryParams)
```
Comment (thc202) [inline zap/client.go]: Should be using `c.Base`?

**Comment by thc202** [comment]
- Could you add a Travis CI configuration file? [1]
 - Would you mind adding an example using the client? (e.g. start the spider, active scan...)


[1] https://docs.travis-ci.com/user/languages/go/

**Comment by mqliang** [comment]
@thc202 I am trying to add an example.

1. start bodgeit at 8090
```
docker run --rm -p 8090:8080 -i -t psiinon/bodgeit
```
2. start zaproxy at 8080
```
docker run -u zap -p 8080:8080 -i owasp/zap2docker-stable zap.sh -daemon -host 0.0.0.0 -port 8080
```
3. write a example using the client
```
package main

import (
	"github.com/zaproxy/zap-api-go/zap"
	"fmt"
	"log"
)

const TARGET = "http://localhost:8090/bodgeit/"

func main() {
	cfg := &zap.Config{
		Proxy: "http://localhost:8080",
	}
	client, err := zap.NewClient(cfg)
	if err != nil {
		log.Fatal(err)
	}
	resp, err := client.Spider().Scan(TARGET, "", "", "", "")
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println(resp)
}
```
4.request was successfully send to zaproxy but unfortunately, zaproxy return an error. And zaproxy's log is really strange, I don't know what's going wrong

```
5023 [ZAP-daemon] INFO org.zaproxy.zap.DaemonBootstrap  - ZAP is now listening on 0.0.0.0:8080
256887 [ZAP-ProxyThread-1] WARN org.zaproxy.zap.extension.api.API  - Request to API URL http://zap/JSON/spider/action/scan/?url=http%3A%2F%2Flocalhost%3A8090%2Fbodgeit%2F from 172.17.0.1 not permitted
```

**Comment by mqliang** [comment]
I try to add a command flag `-config api.addrs.addr.name=127.17.0.1` wen start zaproxy daemon, but from the log it seems the flag was NOT respected. see log:

```
➜  ~ docker run -u zap -p 8080:8080 -i owasp/zap2docker-stable zap.sh -daemon -host 0.0.0.0 -port 8080 -config api.disablekey=true -config api.addrs.addr.name=127.17.0.1
Found Java version 1.8.0_121
Available memory: 1999 MB
Setting jvm heap size: -Xmx499m
265 [main] INFO org.zaproxy.zap.DaemonBootstrap  - OWASP ZAP 2.6.0 started 17/06/17 12:10:29
322 [main] INFO org.parosproxy.paros.common.AbstractParam  - Setting config api.disablekey = true was null
330 [main] INFO org.parosproxy.paros.common.AbstractParam  - Setting config api.addrs.addr.name = 127.17.0.1 was null
```

And docker inspect:
```
➜  ~ docker inspect 9bf46ce8802b
[
    {
        "Id": "9bf46ce8802b4b20b766113daac4e37375e07ee8f9dc7e1a5af29bb1e737095e",
        "Created": "2017-06-17T12:10:28.170912803Z",
        "Path": "zap.sh",
        "Args": [
            "-daemon",
            "-host",
            "0.0.0.0",
            "-port",
            "8080",
            "-config",
            "api.disablekey=true",
            "-config",
            "api.addrs.addr.name=127.17.0.1"
        ],
```

**Comment by mqliang** [comment]
@thc202 I finally solve the `Request to API URL from 172.17.0.1 not permitted` by `-config api.disablekey=true -config api.addrs.addr.name=.\* -config api.addrs.addr.regex=true` flag.

And I just write a example as zap-api-java example does: start the spider, active scan. See my example code here: https://github.com/zaproxy/zap-api-go/pull/2/commits/ccac51bd68e596ca748ccc96aeaaf0120d91f816

However, when the first Spider.Scan() is complete, Ascan.Scan() failed, error log is as follows:
```
4810038 [ZAP-ProxyThread-1508] WARN org.zaproxy.zap.extension.api.API  - ApiException while handling API request:
URL Not Found in the Scan Tree (url_not_found)
	at org.zaproxy.zap.extension.ascan.ActiveScanAPI.scanURL(Unknown Source)
	at org.zaproxy.zap.extension.ascan.ActiveScanAPI.handleApiAction(Unknown Source)
	at org.zaproxy.zap.extension.api.API.handleApiRequest(Unknown Source)
	at org.parosproxy.paros.core.proxy.ProxyThread.processHttp(Unknown Source)
	at org.parosproxy.paros.core.proxy.ProxyThread.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
```

Seems the same issue with https://github.com/zaproxy/zaproxy/issues/3036. You comment under the issue as:

```
You just need to access the URL(s) by/through ZAP (e.g. proxy the request(s), use the spiders) before starting the active scan.
```

However, in my example code, I do use the call Spider.Scan() before starting Ascan.Scan(). Any idea?

**Comment by thc202** [comment]
That's most likely caused by the seed (`TARGET`) passed to the spider, it should use the address of the docker machine.

EDIT: "docker machine" should be "bodgeit container".

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+language: go
+
+go:
+  - 1.6
+  - 1.7
+  - 1.8
+  - tip
+
+services:
+  - zap-zpi-go
```
Comment (thc202) [inline .travis.yml]: There's a typo?

File: pre-commit [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+GOVET_ERRORS=$(go tool vet *.go 2>&1)
+if [ -n "${GOVET_ERRORS}" ]; then
+  printf >&2 'go vet failed for the following reasons:\n%s\n\nplease run "go tool vet *.go" on your changes before committing.\n' "${GOVET_ERRORS}"
+  exit 1
+fi
+
+if [ -z "${NOTEST}" ]; then
+  printf >&2 'Running short tests...\n'
+  env AMQP_URL= go test -short -v | egrep 'PASS|ok'
```
Comment (thc202) [inline pre-commit]: Is the env variable really needed?

**Comment by thc202** [comment]
Travis CI is now enabled for this repo, next push should trigger the build.

**Comment by thc202** [comment]
Regarding https://github.com/zaproxy/zap-api-go/pull/2#issuecomment-309211501, that's the normal output when the config is set. The address has a typo though (i.e. `172.17.0.1` instead of `127.17.0.1`), that's most likely the reason why it didn't work.

**Comment by mqliang** [comment]
@thc202 Done.

NOTE: 
* I disable the golint check in travis. [golint](https://github.com/golang/lint) is the Google coding/commenting style. However, in our case, all comments are extracted from java comments. It will be a nightmare to make them confirm to golint spec.
* I also disable unit test check. Currently we have no unit test.

You can verify this client works by setting up a test environment and run the example code. I have write a guide here https://github.com/zaproxy/zap-api-go/pull/2/commits/6250bfcfe747b614cbba1e6babe779d29e9a545b#diff-20bfc3ff6a860483887b93bf9cf0d135

**Comment by thc202** [comment]
Thanks, will try that.

File: example/example.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2016 The Zed Attack Proxy Team
```
Comment (thc202) [inline example/example.go]: Year should be updated.

File: example/example.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	fmt.Println("Active scan : " + target)
+	resp, err = client.Ascan().Scan(target, "True", "False", "", "", "", "")
+	if err != nil {
+		log.Fatal(err)
+	}
+	// The scan now returns a scan id to support concurrent scanning
+	scanid = resp["scan"].(string)
+	for {
+		time.Sleep(5000 * time.Millisecond)
+		resp, _ = client.Spider().Status(scanid)
```
Comment (thc202) [inline example/example.go]: Should use `Ascan`.

File: example/example.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+		progress, _ := strconv.Atoi(resp["status"].(string))
+		fmt.Printf("Active Scan progress : %d", progress)
```
Comment (thc202) [inline example/example.go]: Needs a newline? Getting `Active Scan progress : 0Active Scan progress : 1`.

File: zap/client.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+	req, err := http.NewRequest("GET", path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(queryParams) == 0 {
+		queryParams = map[string]string{}
+	}
+	// Send the API key even if there are no parameters,
+	// older ZAP versions might need ZPI key as (query) parameter.
```
Comment (thc202) [inline zap/client.go]: There's a typo in `API`.

**Comment by thc202** [comment]
Thanks! Was able to run the example without issues.

Just one question, how can we connect to ZAP using HTTPS? tried to change the proxy URL to use HTTPS, but it still connects with HTTP, does the user needs to change `DefaultBase` for that?

**Comment by mqliang** [comment]
@thc202 HTTPS support is added in my last commit.

File: example/README.md [hunk 8]
```diff
@@ -1,21 +1,46 @@
...
 
 #### run example
-```
+```shell
 go run example.go -target=http://{$bodgeit_IP}:8080/bodgeit
+```
+
+#### How to connect to ZAP server using HTTPS
+
+1. Client commect to ZAP server using HTTP by default. To instruct client to connect ZAP server using HTTPS, you should:
```
Comment (thc202) [inline example/README.md]: There's a typo in "connect" and in "change" (following line).

File: example/README.md [hunk 9]
```diff
@@ -1,21 +1,46 @@
...
+		// you can set your custom certificates here
+		TLSConfig: tls.Config{
+			InsecureSkipVerify: true,
+		},
+	}
+	client, err := zap.NewClient(cfg)
+	if err != nil {
+		log.Fatal(err)
+	}
+    /*
```
Comment (thc202) [inline example/README.md]: Here is using spaces but above lines use tabs. Worth changing, maybe remove indentation?

**Comment by thc202** [comment]
Thanks! Was able to connect to ZAP with HTTPS :)

Other than minor typos, looks good to me.

**Comment by mqliang** [comment]
@thc202 comment addressed. 

As for the space/tab indentation problem, I noticed that in zapproxy project, there are too many inconsistency, even in a single file, some code snippet  use space and some use tab. Maybe it's due to different setting between different editors/IDEs, and there is a awesome project [EditorConfig](http://editorconfig.org/) can help.

**Comment by thc202** [comment]
LGTM, could you squash the commits?

Yeah, we are in the process of normalising that.

**Comment by kingthorin** [comment]
@mqliang was this closed intentionally?

**Comment by mqliang** [comment]
@kingthorin No. I am trying to squash the commits via:
```
git reset --soft <first commit id>
git commit --amend
git push -f origin  client-impl
```
And unfortunately, the pr get closed.

**Comment by mqliang** [comment]
@kingthorin Will open a new pr soon

**Pull request closed without merging**